bumble/hid.py

```diff
@@ -161,31 +171,61 @@ def __bytes__(self) -> bytes:
         return self.header(Message.ControlCommand.VIRTUAL_CABLE_UNPLUG)
 
 
+# Device sends input report, host sends output report.
 @dataclass
 class SendData(Message):
     data: bytes
+    report_type: int
     message_type = Message.MessageType.DATA
 
     def __bytes__(self) -> bytes:
-        return self.header(Message.ReportType.OUTPUT_REPORT) + self.data
+        return self.header(self.report_type) + self.data
+
+
+@dataclass
+class SendHandshakeMessage(Message):
+    result_code: int
+    message_type = Message.MessageType.HANDSHAKE
+
+    def __bytes__(self) -> bytes:
+        return self.header(self.result_code)
 
 
 # -----------------------------------------------------------------------------
-class Host(EventEmitter):
-    l2cap_ctrl_channel: Optional[l2cap.ClassicChannel]
-    l2cap_intr_channel: Optional[l2cap.ClassicChannel]
+class HID(EventEmitter):
+    l2cap_ctrl_channel: Optional[l2cap.ClassicChannel] = None
+    l2cap_intr_channel: Optional[l2cap.ClassicChannel] = None
+    connection: Optional[device.Connection] = None
+
+    class Role(enum.IntEnum):
+        HOST = 0x00
+        DEVICE = 0x01
 
-    def __init__(self, device: Device, connection: Connection) -> None:
+    def __init__(self, device: device.Device, role: Role) -> None:
         super().__init__()
         self.device = device
-        self.connection = connection
-
-        self.l2cap_ctrl_channel = None
-        self.l2cap_intr_channel = None
+        self.role = role
 
         # Register ourselves with the L2CAP channel manager
-        device.register_l2cap_server(HID_CONTROL_PSM, self.on_connection)
-        device.register_l2cap_server(HID_INTERRUPT_PSM, self.on_connection)
+        device.register_l2cap_server(HID_CONTROL_PSM, self.on_l2cap_connection)
+        device.register_l2cap_server(HID_INTERRUPT_PSM, self.on_l2cap_connection)
+
+        device.on('connection', self.on_device_connection)
+
+    def handle_get_report(self, pdu: bytes):
+        return
+
+    def handle_set_report(self, pdu: bytes):
+        return
+
+    def handle_get_protocol(self, pdu: bytes):
+        return
+
+    def handle_set_protocol(self, pdu: bytes):
+        return
+
+    def send_handshake_message(self, result_code: int):
+        return
```

Comment:
What's the usage of these empty methods?
If that's a TODO item, please add TODO annotations.

Reply:
These are override functions which are defined in Device class and called in HID class

Reply:
In that case, it's better to mark this class `abc.ABC` and mark these methods `@abc.abstractmethod`

https://docs.python.org/3/library/abc.html#abc.abstractmethod

Reply:
Also adding `@typing_extensions.override` in the derived methods.

https://typing-extensions.readthedocs.io/en/latest/#override

Reply:
Thank you Josh. We have updated the code with abstract class and method.

Regards,
LTIMindtree Team